File: 3398.smallest-substring-with-identical-characters-i.py
#
# @lc app=leetcode id=3398 lang=python3
#
# [3398] Smallest Substring With Identical Characters I
#

# @lc code=start
class Solution:
    def minLength(self, s: str, numOps: int) -> int:
        n = len(s)
        if n == 0: return 0
        # Prefix sums for zeros and ones
        prefix_zeros = [0] * (n + 1)
        prefix_ones = [0] * (n + 1)
        for i in range(n):
            prefix_zeros[i+1] = prefix_zeros[i] + (1 if s[i] == '0' else 0)
            prefix_ones[i+1] = prefix_ones[i] + (1 if s[i] == '1' else 0)
        min_length = n
        # Sliding window approach to use flips effectively
        for start in range(n):
            for end in range(start, n):
                zeros_in_window = prefix_zeros[end+1] - prefix_zeros[start]
                ones_in_window = prefix_ones[end+1] - prefix_ones[start]
                flips_needed = min(zeros_in_window, ones_in_window)
                if flips_needed <= numOps:
                    min_length = min(min_length, end - start + 1)
        return min_length
# @lc code=end

File: 3398.smallest-substring-with-identical-characters-i.cpp
#
# @lc app=leetcode id=3398 lang=cpp
#
# [3398] Smallest Substring With Identical Characters I
#

# @lc code=start
class Solution {
public:
    int minLength(string s, int numOps) {
        // Initialize variables to track left and right pointers, number of zeroes and ones.
        int n = s.size();
        int left = 0, right = 0;
        int countZero = 0, countOne = 0;
        int maxBlockSize = INT_MAX;
        
        // Use sliding window technique.
        while (right < n) {
            // Increment counts based on current character.
            if (s[right] == '0') {
                countZero++;
            } else {
                countOne++;
            }
            
            // If more than numOps flips needed for either character, move left pointer.
            while (std::min(countZero, countOne) > numOps) {
                if (s[left] == '0') {
                    countZero--;
                } else {
                    countOne--;
                }
                left++;
            }
            
            // Update maxBlockSize with current window size.
            maxBlockSize = std::min(maxBlockSize, right - left + 1);
            right++;
        } 
        return maxBlockSize;   	// Return minimum block size found after operations. 	}	}; # @lc code=end